datalad/cmd.py

```diff
@@ -194,30 +201,39 @@ def run(self, cmd, log_stdout=True, log_stderr=True, log_online=False,
 
         self.log("Running: %s" % (cmd,))
 
-        if not self.dry:
+        if self.protocol.do_execute():
 
             if shell is None:
                 shell = isinstance(cmd, basestring)
 
-            if self.protocol:
-                self.protocol.write_section(
-                    shlex.split(cmd) if isinstance(cmd, basestring) else cmd)
+            if self.protocol.write_ext_commands():
+                prot_exc = None
+                t_start = time.time()
+
             try:
                 proc = subprocess.Popen(cmd, stdout=outputstream,
                                         stderr=errstream,
                                         shell=shell,
                                         cwd=cwd or self.cwd,
                                         env=env or self.env)
+                # shell=True allows for piping, multiple commands, etc.,
+                # but that implies to not use shlex.split()
+                # and is considered to be a security hazard.
+                # So be careful with input.
+                # Alternatively we would have to parse `cmd` and create
+                # multiple subprocesses.
             except Exception, e:
+                prot_exc = e
                 lgr.error("Failed to start %r%r: %s" %
                           (cmd, " under %r" % cwd if cwd else '', e))
                 raise
-            # shell=True allows for piping, multiple commands, etc.,
-            # but that implies to not use shlex.split()
-            # and is considered to be a security hazard.
-            # So be careful with input.
-            # Alternatively we would have to parse `cmd` and create multiple
-            # subprocesses.
+
+            finally:
+                if self.protocol.write_ext_commands():
+                    prot_time = time.time() - t_start
```

Comment:
I feel that this should be implemented by the protocol itself  ...  then may be it needs "start_section" , "end_section" and may be protocol should expose itself also as a context manager